apps/networkmonitor/networkmonitor.nim

```diff
@@ -82,43 +99,68 @@ proc setConnectedPeersMetrics(discoveredNodes: seq[Node],
     let ip = $typedRecord.get().ip.get().join(".")
     allPeers[peerId].ip = ip
 
-    let peer = toRemotePeerInfo(discNode.record)
-    if not peer.isOk():
-      warn "error converting record to remote peer info", record=discNode.record
-      continue
+    # try to ping the peer
+    if getTime().toUnix() >= allPeers[peerId].lastTimeConnected + ReconnectTime and allPeers[peerId].retries < MaxConnectionRetries:
+      if allPeers[peerId].retries > 0:
+        warn "trying to dial failed peer again", peerId=peerId, retry=allPeers[peerId].retries
 
-    # try to connect to the peer
-    # TODO: check last connection time and if not > x, skip connecting
-    let timedOut = not await node.connectToNodes(@[peer.get()]).withTimeout(timeout)
-    if timedOut:
-      warn "could not connect to peer, timedout", timeout=timeout, peer=peer.get()
-      # TODO: Add other staates
-      allPeers[peerId].connError = "timedout"
-      continue
+      var pingDelay:chronos.Duration
+
+      proc ping(): Future[Result[void, string]] {.async, gcsafe.} =
+        try:
+          let conn = await node.switch.dial(peerInfo.peerId, peerInfo.addrs, PingCodec)
+          pingDelay = await node.libp2pPing.ping(conn)
+          return ok()
+
+        except CatchableError:
+          var msg = getCurrentExceptionMsg()
+          if msg == "Future operation cancelled!":
+            msg = "timedout"
+          warn "failed to ping the peer", peer=peerInfo, err=msg
+
+          allPeers[peerId].connError = msg
+          return err("could not ping peer: " & msg)
+
+      let timedOut = not await ping().withTimeout(timeout)
+      # need this check for pingDelat == 0 because there may be a conn error before timeout
+      if timedOut or pingDelay == 0.millis:
+        allPeers[peerId].retries += 1
+        continue
+
+      info "successfully pinged peer", peer=peerInfo, duration=pingDelay.millis
+      peer_ping.observe(pingDelay.millis)
+
+      if allPeers[peerId].avgPingDuration == 0.millis:
+        allPeers[peerId].avgPingDuration = pingDelay
+
+      # TODO: check why the calculation ends up losing precision
+      allPeers[peerId].avgPingDuration = int64((float64(allPeers[peerId].avgPingDuration.millis) * (AvgPingWindow - 1.0) + float64(pingDelay.millis)) / AvgPingWindow).millis
```

Comment:
is this formula correct?
think its missing a `/AvgPingWindow`
old average * (n-1)/n + new value /n

something like this (unsure about the castings)
```
allPeers[peerId].avgPingDuration = int64((float64(allPeers[peerId].avgPingDuration.millis) * (AvgPingWindow - 1.0)/AvgPingWindow + float64(pingDelay.millis)) / AvgPingWindow).millis
```

Reply:
Hmm.. You produced exactly the same formula if I see correctly:D

Reply:
No, that would not be correct IMO - you need to make avg represent 9 values in the window + new value, what you have there is make avg*0.9 + new value and then divide by 10 (check the parentheses)

I believe my formula is correct, just some weird issue with casting/rounding

Reply:
missread the parenthesis in your formula. yours and mine are the same.

> what you have there is make avg*0.9 + new value and then divide by 10 

nope, since division has priority over sum, 
`old_average * (n-1)/n + new_value /n`

is equivalent to

``old_average * (n-1)/n + (new_value /n)``